zyte_spider_templates/spiders/serp.py

```diff
@@ -117,12 +136,30 @@ def start_requests(self) -> Iterable[Request]:
         url = f"https://www.{self.args.domain.value}/search"
         for search_query in search_queries:
             search_url = add_or_replace_parameter(url, "q", search_query)
-            for start in range(0, self.args.max_pages * 10, 10):
-                if start:
-                    search_url = add_or_replace_parameter(
-                        search_url, "start", str(start)
-                    )
-                yield self.get_start_request(search_url)
-
-    def parse_serp(self, response) -> Iterable[Serp]:
-        yield Serp.from_dict(response.raw_api_response["serp"])
+            yield self.get_serp_request(search_url, page_number=1)
+
+    def parse_serp(self, response, page_number=_UNSET) -> Iterable[Serp]:
+        serp = Serp.from_dict(response.raw_api_response["serp"])
+
+        if page_number is not _UNSET:
+            next_start = page_number * self._results_per_page
+            if serp.organicResults and serp.metadata.totalOrganicResults > next_start:
+                next_url = add_or_replace_parameter(serp.url, "start", str(next_start))
+                yield self.get_serp_request(next_url, page_number=page_number + 1)
+        else:
+            warn(
+                (
+                    "zyte_spider_templates.spiders.serp.GoogleSearchSpider.parse_serp "
+                    "now expects a page_number callback keyword argument "
+                    "(cb_kwargs) indicating the page being parsed (e.g. 1 for "
+                    "the 1st page), and will yield a request for the next "
+                    "page if the next page could have additional results, "
+                    "instead of expecting the source callback to take care of "
+                    "pagination by yielding requests for multiple pages in "
+                    "parallel."
+                ),
+                DeprecationWarning,
+                stacklevel=2,
+            )
```

Comment:
It could be fine to drop all the warnings as well :) It's nice to have the warnings, but it seems we don't have many users, and it complicates the code, so it may be better to just do a backwards-incompatible change for now.